src/stories/components/RecipeReviews/RecipeReviews.js

```diff
@@ -17,58 +22,166 @@ import { FormItem } from '../FormItem/FormItem';
+  // to leave username as part of review
+  const { currentUser } = useContext(AuthContext);
+  const [profile, setProfile] = useState(null);
+  const history = useHistory();
+
+  const [errorMessage, setErrorMessage] = useState('');
+  const [successMessage, setSuccessMessage] = useState('');
   const [reviews, setReviews] = useState(null);
   const [review, setReview] = useState({
-    recipeId,
+    // recipeId,
     title: '',
     content: '',
     rating: '',
-    date: (new Date).toLocaleDateString('default', {
-      month: 'long',
-      day: 'numeric',
-      year: 'numeric',
-    }),
-    user: 'apizzoccheri',
+    username: '',
+    date: UTILS.convertDateToString(new Date()),
   });
 
   useEffect(() => {
-    // TODO make call to DB to fecth the recipe's reviews
-    setReviews([{
-      title: 'Test',
-      content: 'Hello world',
-      rating: '4',
-      username: 'apizzoccheri',
-      date: new Date(),
-    }]);
-  }, [reviews]);
-
-  // push review to database
+    setErrorMessage('');
+    setSuccessMessage('');
+    // get profiles username to attach to reviews
+    if (currentUser) {
+      database.getProfile(currentUser.uid)
+        .then((doc) => {
+          setProfile(doc.data());
+        })
+        .catch((error) => {
+          // setError(error)
+        });
+    }
+
+    // this call is purely to make document discoverable in future
+    database.setRecipe(recipeId);
+    // this one actually gets data
+    setReviewsFromDB();
+  }, [currentUser]);
+
+
+  /**
+   * Makes call to Firestore, grabs all reviews for this recipe
+   *   and saves them to the local state variable 'reviews'
+   */
+  const setReviewsFromDB = () => {
+    setReviews([]);
+    database.getReviewsForRecipe(recipeId)
+      .then((collection) => {
+        const theReviews = [];
+        collection.forEach((doc) => {
+          theReviews.push(doc.data());
+        });
+        setReviews(theReviews);
+      });
+  };
+
+  /**
+   * Validates for not-null options in form, if all not null
+   *   submits the current review to the database, clears
+   *   the fields, and then re-grabs all the reviews for the
+   *   current recipe
+   * Redirects to login if no current user.
+   */
+  const checkAndSubmitReviewToDB = () => {
+    setErrorMessage('');
+    setSuccessMessage('');
+
+    // basic data validation, check all for non-empty
+    if ( profile === null ) {
+      // this first if is JANK but getting profile back takes forever
+      //   gotta give time for the promise to resolve before user can
+      //   submit a review
+      setErrorMessage( 'Please try again in a second.' );
+    } else if ( review.title === '' ) {
+      setErrorMessage( 'Please enter a review title.' );
+    } else if ( review.content === '' ) {
+      setErrorMessage('Please enter content of the review.' );
+    } else if ( review.rating === '' ) {
+      setErrorMessage('Please give a rating for this review.' );
```

Comment:
Sorry not quite understanding, should I add required to the Form component itself? Or is there some way I can do it from up in RecipeReview. 

Seems to make sense that all FormItems might always be required, given that they're consciously part of the form? So possibly could toss it in the class itself. Ive been reading through documentation for 'required' as part of a Form and haven't been able to get that to catch empty fields. 

I get that it makes sense to have validation somewhere else maybe, so that that code doesn't have to be repeated. I'm just not getting how to do that validation easily by having the fields be required.

Reply:
Gonna put a pause on this and will circle back as we do clean up.